### PR TITLE
Add automation to auto publish builds to Bintray from CI

### PR DESCRIPTION
**Pull request by mokagio**
This uses the work from #39 to run the `bintrayPublish` task from CI and allow us to have development builds for every push to a feature branch, as well as beta and release builds, on Bintray.

### Workflow choices

If we try to upload an artifact for a version that is already available on the repository, the API will return a 409 error. @loremattei and I discussed this, and decided to ship the first version of this automation under the assumption that all merges in `develop` and `trunk` should have a version bump.

I have a new Peril rule that [I'm working on](https://github.com/Automattic/peril-settings/pull/58) to fail a PR to a main branch that doesn't update the library's version.

## To Test

You can see that this works by comparing the [output on CI](https://app.circleci.com/pipelines/github/wordpress-mobile/WordPress-Utils-Android/140/workflows/e1e88b98-4c59-4061-b3be-0021e1822440/jobs/267) with the activity log in [the Bintray page](https://bintray.com/wordpress-mobile/).

CI | Bintray
--- | ---
image | image

I opened [this demo PR in my fork](https://github.com/mokagio/WordPress-Utils-Android/pull/10) to show that the build works when merging on `develop`, see the links in the description.

[This failed build](https://app.circleci.com/pipelines/github/mokagio/WordPress-Utils-Android/35/workflows/c85976eb-1f6e-4648-a6e4-46582b39e224) shows what happens when trying to upload an artifact for a version that already exist.

To test this for yourself, you can follow the same approach I used:

- Create your own Maven repository on Bintray
- Fork this repo
- Create a branch based off this one to your fork
- Point the Bintray plugin to your fork (see [this](https://github.com/mokagio/WordPress-Utils-Android/pull/10/commits/7d5db1505fa429e184c5ba63f89735def44182de))
- Push the branch and open a PR against your fork

### CODE REVIEW

File: .circleci/config.yml
Comment (mokagio) [inline .circleci/config.yml:69]: If a build runs on a branch with no PR, then we wouldn't be able to generate the `PR-SHA` version.

This can happen when you push a branch to the remote but don't open the PR straight afterwards.

If you then open a PR and run the build again, CircleCI is smart enough to realize that there is now a PR for that commit and provide the environment variable.

[Here's](https://app.circleci.com/pipelines/github/mokagio/WordPress-Utils-Android/32/workflows/3fdea926-919f-40f8-b17b-9e4bdc46f5ec/jobs/42) an example of this behavior in action, I pushed a branch to my fork and did not open a PR for it

image

And this shows that when I re-run the build, the upload worked. Notice the difference in time between the two publishing jobs

image

As a future refinement, I think it would be good to cancel the job here, just to make it clearer to the developers that no artifact was published.

File: .circleci/config.yml
Comment (mokagio) [inline .circleci/config.yml]: My understanding of Gradle is pretty superficial. If you know why a task would behave so differently between CI and local, and have a suggestion on how to avoid this long winded workaround, I'd love to know about it.

The commits in #40 show the different things I tried before coming to this point, including builds failing without finding the artifact.

File: .circleci/config.yml
Comment (loremattei) [inline .circleci/config.yml:91]: I'm not sure the problem it mentions there is related 🤔

File: .circleci/config.yml
Comment (loremattei) [reply .circleci/config.yml]: This is weird. I wonder if it can be related to the different OS (OSX vs Linux). 
Anyway, I don't dislike that we have full control on what and how we build before uploading it tbh.

**Comment by loremattei** [comment]
This looks good to me. Since we'll want this same behaviour in many different repositories, should we move it to a shared orb?

File: .circleci/config.yml
Comment (mchowning) [reply .circleci/config.yml]: That is really weird that the task dependencies would be different on CI. 🤔  It might be interesting to see what the output of `./gradlew bintrayUpload --dry-run` looks like on CI compared to what that command outputs locally.

Comment (mokagio) [reply .circleci/config.yml]: @mchowning good question 🤔 

I tried this out [in this build](https://app.circleci.com/pipelines/github/wordpress-mobile/WordPress-Utils-Android/150/workflows/a54c8976-68ee-41b8-8446-6bccdc89c7d2/jobs/298) via #52.

### CI output

```
$ ./gradlew bintrayUpload --dry-run -PbintrayVersion=test-dry-run-1

Starting a Gradle Daemon, 1 incompatible and 1 stopped Daemons could not be reused, use --status for details
Configuration on demand is an incubating feature.

> Configure project :WordPressUtils
WARNING: API 'variant.getJavaCompile()' is obsolete and has been replaced with 'variant.getJavaCompileProvider()'.
It will be removed in version 5.0 of the Android Gradle plugin.
For more information, see https://d.android.com/r/tools/task-configuration-avoidance.
To determine what is calling variant.getJavaCompile(), use -Pandroid.debug.obsoleteApi=true on the command line to display more information.
:WordPressUtils:bintrayUpload SKIPPED
:bintrayUpload SKIPPED
:bintrayPublish SKIPPED

BUILD SUCCESSFUL in 6s
```

### Local output

```
➜ ./gradlew bintrayUpload --dry-run -PbintrayVersion=test-dry-run-1  

> Configure project :WordPressUtils
WARNING: API 'variant.getJavaCompile()' is obsolete and has been replaced with 'variant.getJavaCompileProvider()'.
It will be removed in version 5.0 of the Android Gradle plugin.
For more information, see https://d.android.com/r/tools/task-configuration-avoidance.
To determine what is calling variant.getJavaCompile(), use -Pandroid.debug.obsoleteApi=true on the command line to display more information.
:WordPressUtils:preBuild SKIPPED
:WordPressUtils:preReleaseBuild SKIPPED
:WordPressUtils:compileReleaseAidl SKIPPED
:WordPressUtils:mergeReleaseJniLibFolders SKIPPED
:WordPressUtils:mergeReleaseNativeLibs SKIPPED
:WordPressUtils:stripReleaseDebugSymbols SKIPPED
:WordPressUtils:copyReleaseJniLibsProjectAndLocalJars SKIPPED
:WordPressUtils:compileReleaseRenderscript SKIPPED
:WordPressUtils:generateReleaseBuildConfig SKIPPED
:WordPressUtils:generateReleaseResValues SKIPPED
:WordPressUtils:generateReleaseResources SKIPPED
:WordPressUtils:packageReleaseResources SKIPPED
:WordPressUtils:parseReleaseLocalResources SKIPPED
:WordPressUtils:processReleaseManifest SKIPPED
:WordPressUtils:generateReleaseRFile SKIPPED
:WordPressUtils:compileReleaseKotlin SKIPPED
:WordPressUtils:javaPreCompileRelease SKIPPED
:WordPressUtils:compileReleaseJavaWithJavac SKIPPED
:WordPressUtils:extractReleaseAnnotations SKIPPED
:WordPressUtils:mergeReleaseGeneratedProguardFiles SKIPPED
:WordPressUtils:mergeReleaseConsumerProguardFiles SKIPPED
:WordPressUtils:mergeReleaseShaders SKIPPED
:WordPressUtils:compileReleaseShaders SKIPPED
:WordPressUtils:generateReleaseAssets SKIPPED
:WordPressUtils:packageReleaseAssets SKIPPED
:WordPressUtils:packageReleaseRenderscript SKIPPED
:WordPressUtils:prepareLintJarForPublish SKIPPED
:WordPressUtils:processReleaseJavaRes SKIPPED
:WordPressUtils:mergeReleaseJavaResource SKIPPED
:WordPressUtils:syncReleaseLibJars SKIPPED
:WordPressUtils:bundleReleaseAar SKIPPED
:WordPressUtils:generateMetadataFileForUtilsPublicationPublication SKIPPED
:WordPressUtils:generatePomFileForUtilsPublicationPublication SKIPPED
:WordPressUtils:publishUtilsPublicationPublicationToMavenLocal SKIPPED
:WordPressUtils:bintrayUpload SKIPPED
:bintrayUpload SKIPPED
:bintrayPublish SKIPPED

BUILD SUCCESSFUL in 743ms
```

There's quite a difference 😳

File: .circleci/config.yml
Comment (mokagio) [reply .circleci/config.yml:91]: Good point... That documentation only points out that, depending on what you are trying to do, e.g. include the dependency chain in the POM, you might need to add some workaround.

I'd like to leave the comment in there, to help future us find information that might be helpful in the future. I'll reword it.

**Comment by mokagio** [comment]
> Since we'll want this same behaviour in many different repositories, should we move it to a shared orb?

Thank you @loremattei.

I'll see if I get the time to extract this in the orb. Would it be okay to merge this at it is as a starting point, so that we can start seeing out it behaves in the real world?

Tweaking the CircleCI configuration would be easier if it's hardcoded here rather than with the extra step of submitting changes to the orb and shipping new versions. Maybe we could make the switch when it's time to apply this to another project?

File: .circleci/config.yml
Comment (mchowning) [reply .circleci/config.yml]: > ...There's quite a difference 😳

For sure! That is so weird. I do not have a good idea why this wouldn't be working. Maybe the `WordPressUtils:publishUtilsPublicationPublicationToMavenLocal` doesn't get created on CI for some reason and that somehow breaks the task dependency chain? That's really just wild speculation on my part to be honest though. I'm a bit stumped.

You may have already tried this, but I'm curious what happens if you explicitly make the `bintrayPublish` task depend on `generatePomFileForUtilsPublicationPublication`, and then, if necessary, also explicitly make `generatePomFileForUtilsPublicationPublication` depend on `assembleRelease`? It doesn't seem like we should _have_ to do that, but it might fix the problem.

FWIW, what you have here seems fine, so I wouldn't disagree if you felt like it wasn't worth spending more time figuring this out. 

Even if we have to be explicit about invoking each task, It would be nice if we could combine them into a single build with something like `./gradlew assembleRelease generatePomFileForUtilsPublicationPublication bintrayUpload`. That would speed up the builds by not re-doing work (although caching reduces the re-work, I believe that at least the configuration phase will get re-run each time). BUT, given that the task dependencies clearly aren't working the way we expect on CI, I'm not feeling very confident that the tasks would always be run in the right order if we combined them into one build command. 😞

**Comment by loremattei** [comment]
I've had a bit more time to look into it today and I figured out what's going on. 
Running a build scan showed that we have the `configure-on-demand` feature enabled on CI and looking at our configuration I found that [we enable it in our orb](https://github.com/wordpress-mobile/circleci-orbs/blob/82f9821130890756e85277486ca0dde6cae25431/src/android/orb.yml#L19). 

I then tried to run a `dry-run` on CI with that feature disabled (`./gradlew --dry-run bintrayUpload --no-configure-on-demand`) and it now looks pretty good: 

```
WARNING: API 'variant.getJavaCompile()' is obsolete and has been replaced with 'variant.getJavaCompileProvider()'.
It will be removed in version 5.0 of the Android Gradle plugin.
For more information, see https://d.android.com/r/tools/task-configuration-avoidance.
To determine what is calling variant.getJavaCompile(), use -Pandroid.debug.obsoleteApi=true on the command line to display more information.
:WordPressUtils:preBuild SKIPPED
:WordPressUtils:preReleaseBuild SKIPPED
:WordPressUtils:compileReleaseAidl SKIPPED
:WordPressUtils:mergeReleaseJniLibFolders SKIPPED
:WordPressUtils:mergeReleaseNativeLibs SKIPPED
:WordPressUtils:stripReleaseDebugSymbols SKIPPED
:WordPressUtils:copyReleaseJniLibsProjectAndLocalJars SKIPPED
:WordPressUtils:compileReleaseRenderscript SKIPPED
:WordPressUtils:generateReleaseBuildConfig SKIPPED
:WordPressUtils:generateReleaseResValues SKIPPED
:WordPressUtils:generateReleaseResources SKIPPED
:WordPressUtils:packageReleaseResources SKIPPED
:WordPressUtils:parseReleaseLocalResources SKIPPED
:WordPressUtils:processReleaseManifest SKIPPED
:WordPressUtils:generateReleaseRFile SKIPPED
:WordPressUtils:compileReleaseKotlin SKIPPED
:WordPressUtils:javaPreCompileRelease SKIPPED
:WordPressUtils:compileReleaseJavaWithJavac SKIPPED
:WordPressUtils:extractReleaseAnnotations SKIPPED
:WordPressUtils:mergeReleaseGeneratedProguardFiles SKIPPED
:WordPressUtils:mergeReleaseConsumerProguardFiles SKIPPED
:WordPressUtils:mergeReleaseShaders SKIPPED
:WordPressUtils:compileReleaseShaders SKIPPED
:WordPressUtils:generateReleaseAssets SKIPPED
:WordPressUtils:packageReleaseAssets SKIPPED
:WordPressUtils:packageReleaseRenderscript SKIPPED
:WordPressUtils:prepareLintJarForPublish SKIPPED
:WordPressUtils:processReleaseJavaRes SKIPPED
:WordPressUtils:mergeReleaseJavaResource SKIPPED
:WordPressUtils:syncReleaseLibJars SKIPPED
:WordPressUtils:bundleReleaseAar SKIPPED
:WordPressUtils:generateMetadataFileForUtilsPublicationPublication SKIPPED
:WordPressUtils:generatePomFileForUtilsPublicationPublication SKIPPED
:WordPressUtils:publishUtilsPublicationPublicationToMavenLocal SKIPPED
:WordPressUtils:bintrayUpload SKIPPED
:bintrayUpload SKIPPED
:bintrayPublish SKIPPED
```

When we review our orbs, we may want to remove it from our configuration, though it will probably impact performance negatively in other projects, so I think our best bet would be to make it optional, keeping the default as is, and giving projects the chance to disable it. 

Right now, you can override it manually when invoking this task. 

Also, I agree with @mchowning that I'm fine to merge this as is if you don't have to update it before your AFK and iterate on this with a separate PR. We're going to port this setup at least to Gutenberg mobile soon, so we'll have chances to improve it.

**Comment by mokagio** [comment]
Thanks @loremattei & @mchowning for reviewing and taking the time to explore different options. All this is great learning for me. 🙇‍♂️ 

> Also, I agree with @mchowning that I'm fine to merge this as is if you don't have to update it before your AFK and iterate on this with a separate PR. 

That seems like the best course of action. I'll merge this now and make a note to work on your improvements suggestion once I'm back, in two weeks time.

**Comment by mokagio** [comment]
> I'll merge this now...

I can't merge it because there's no explicit ✅ and I don't want to disable the rule in the settings and risk forgetting about re-activating it back

**Review by aforcier — APPROVED** [review]
Just got around to taking a look at this - it makes sense to me, though I'd like to see some README updates/documentation on how this process should work. We should probably also update the automated message Peril shows on all WPAndroid PRs that make changes to the subtree.

Nothing to add to what @loremattei and @mchowning already mentioned, sounds like another iteration will happen to resolve the `configure-on-demand` issue.

I'll approve since it's needed to merge, in their stead 😀

**Comment by mokagio** [comment]
Thank you @aforcier!

**Pull request merged**
